examples/input-timestamps-unstable-v1.c

```diff
@@ -0,0 +1,565 @@
+	.keymap = keyboard_handle_keymap,
+	.enter = keyboard_handle_enter,
+	.leave = keyboard_handle_leave,
+	.key = keyboard_handle_key,
+	.modifiers = keyboard_handle_modifiers,
+	.repeat_info = keyboard_handle_repeat_info,
+};
+
+
+/**
+ * xdg_surface handling
+ *
+ * We do nothing at all.
+ */
+
+static void xdg_surface_handle_configure(void *data,
+		struct xdg_surface *xdg_surface, uint32_t serial) {
+	xdg_surface_ack_configure(xdg_surface, serial);
+}
+
+static const struct xdg_surface_listener xdg_surface_listener = {
+	.configure = xdg_surface_handle_configure,
+};
+
+
+/**
+ * xdg_toplevel handling
+ *
+ * We do nothing at all, not even exiting gracefully.
+ */
+
+static void xdg_toplevel_handle_configure(void *data,
+		struct xdg_toplevel *xdg_toplevel, int32_t width, int32_t height,
+		struct wl_array *states) {
+	/* This space intentionally left blank */
+}
+
+static void xdg_toplevel_handle_close(void *data,
+		struct xdg_toplevel *xdg_toplevel) {
+	exit(EXIT_SUCCESS);
+}
+
+static const struct xdg_toplevel_listener xdg_toplevel_listener = {
+	.configure = xdg_toplevel_handle_configure,
+	.close = xdg_toplevel_handle_close,
+};
+
+
+/**
+ * input_timestamps handling
+ *
+ * We simply store the received timestamp in the timespec struct pointed to by
+ * the user data.
+ */
+
+static void input_timestamps_handle_timestamp(void *data,
+		struct zwp_input_timestamps_v1 *zwp_input_timestamps_v1,
+		uint32_t tv_sec_hi, uint32_t tv_sec_lo, uint32_t tv_nsec) {
+	struct timespec *timestamp = data;
+
+	timestamp->tv_sec = ((uint64_t) tv_sec_hi << 32) + tv_sec_lo;
+	timestamp->tv_nsec = tv_nsec;
+}
+
+static const struct zwp_input_timestamps_v1_listener input_timestamps_listener = {
+	.timestamp = input_timestamps_handle_timestamp,
+};
+
+
+/**
+ * helper functions:
+ *
+ * create_anonymous_file():
+ * Helper function from Weston. Creates a new, unique, anonymous file of the
+ * given size. See "os_create_anonymous_file()" for more details.
+ *
+ * create_buffer():
+ * Allocates shm and attaches it to a buffer.
+ *
+ * paint_pixels():
+ * Sets all pixels of the buffer to a fixed color.
+ *
+ * update_current_time():
+ * Updates the global clocks with two samples.
+ *
+ * print_time_and_latency():
+ * Prints the input event timestamps, along with estimated latency measured by
+ * the previous function.
+ */
+
+int create_anonymous_file(off_t size) {
+	const char template[] = "/input-timestamps-example-XXXXXX";
+	const char *path;
+	char *name;
+	int fd;
+
+	path = getenv("XDG_RUNTIME_DIR");
+	if (!path) {
+		return -1;
+	}
+
+	name = malloc(strlen(path) + sizeof(template));
+	if (!name) {
+		return -1;
+	}
+
+	strcpy(name, path);
+	strcat(name, template);
+
+	fd = mkstemp(name);
+	if (fd < 0) {
+		free(name);
+		return -1;
+	}
+
+	unlink(name);
+	free(name);
+
+	if (fcntl(fd, F_SETFD, fcntl(fd, F_GETFD) | FD_CLOEXEC) == -1) {
+		close(fd);
+		return -1;
+	}
+
+	if (ftruncate(fd, size) < 0) {
+		close(fd);
+		return -1;
+	}
+
+	return fd;
+}
```

Comment:
We have a helper elsewhere for this stuff. It would probably also be simpler/shorter to use a DRM surface instead of an SHM one.

Reply:
You probably mean GL surface